monadic-rx/shared/src/main/scala/mhtml/rx.scala

```diff
@@ -142,7 +142,18 @@ sealed trait Rx[+A] { self =>
    */
   def sampleOn[B](other: Rx[B]): Rx[A] = SampleOn[A, B](this, other)
 
+  /**
+    * Memoizes this `Rx` using an internal `Var`. This is primarily
+    * useful for optimizing an Rx graph, so that this `Rx` is only
+    * computed once.
+    */
+  def sharing: Rx[A] = Sharing[A](this)
+
   val impure: RxImpureOps[A] = RxImpureOps[A](this)
+
+  protected var isSharing = false
+  protected var sharingMemo: Option[Any] = None
```

Comment:
`Option[A]`?

Reply:
I seem to get a variance error if I do anything like that, which I guess makes sense being that Rx isn't invariant.

Reply:
I remember seeing that as well, do you also get the error if you move the field to the subclass?

Reply:
Oh, that works! Is this because the case classes extending `Rx` are actually invariant in `A`?

Reply:
@OlivierBlanvillain Well, upon compile, I do see the following for `shareRx.sharingMemo = Option(x)`:

```
[error] /home/brandon/workspace/monadic-html/monadic-rx/shared/src/main/scala/mhtml/rx.scala:268: type mismatch;
[error]  found   : x.type (with underlying type A)
[error]  required: ?A10 where type ?A10 <: A (this is a GADT skolem)
[error]           shareRx.sharingMemo = Option(x)
```

I looked into this a bit but am possibly too sleep deprived and/or lacking in knowledge to know if an immediate solution (even seeing your prior comment about a similar error in `Imitate`.

When I tried out:

```scala
    case shareRxGADT @ Sharing(self) =>
      val shareRx: Sharing[A] = shareRxGADT
      if (!shareRx.isSharing) {
```

```
[error]  found   : mhtml.Rx.Sharing[?A10] where type ?A10 <: A (this is a GADT skolem)
[error]  required: mhtml.Rx.Sharing[A]
[error] Note: ?A10 <: A, but class Sharing is invariant in type A.
[error] You may wish to define A as +A instead. (SLS 4.5)
[error]       val shareRx: Sharing[A] = shareRxGADT
[error]                                 ^
[error] one error found
```

So it looks like we could avoid this error, but we'll be back to using `Option[Any]`. I pushed what I have for now.

Reply:
This looks like a pattern matching bug. Given a `Rx[A]` that is an instance of `Sharing[A](self)` it's perfectly valid to put a `Some[A]` inside the sharing. But somehow `scalac` doesn't know than these two `A`s are the same...

Reply:
When I wrote [these lines](https://github.com/OlivierBlanvillain/monadic-html/blob/master/monadic-rx/shared/src/main/scala/mhtml/rx.scala#L218-L219) I remember trying it out in Dotty, which does the correct thing in this case :smile: